File: Examples/Pascal/CANdoAUTO/IEEE754.pas
//--------------------------------------------------------------------------
//--------------------------------------------------------------------------
// TITLE : IEEE754 floating point number routines.
//
// DESCRIPTION : Routines to deal with encoding/decoding of floating point
//               numbers in a binary format.
//
// DATE : 27/05/14
//
// MODIFICATIONS :
//    30/06/13 Created
//    27/05/14 CharInSet function added for FPC
//
// LICENSE :
//  The SDK (Software Development Kit) provided for use with the CANdo
//  device is issued as FREE software, meaning that it is free for personal,
//  educational & commercial use, without restriction or time limit. The
//  software is supplied "as is", with no implied warranties or guarantees.
//
// AUTHOR : Martyn Brown
//
// © 2013-14 Netronics Ltd. All rights reserved.
//--------------------------------------------------------------------------
//--------------------------------------------------------------------------
unit IEEE754;

interface

uses
  SysUtils, Math;

  function FloatToIEEE754(Number : Double) : Integer;
  function IEEE754ToFloat(Number : Integer) : Double;
  function CheckFloatingPointNumber(Number : String) : Boolean;

implementation

type
TCharSet = set of Char;

{$IFDEF FPC}
function CharInSet(const Character : Char; const ASet : TCharSet): Boolean;
begin
  Result := Character in ASet;
end;
{$ENDIF}
//--------------------------------------------------------------------------
// FloatToIEEE754
//
// Encode a floating point number according to the truncated 24 bit IEEE754
// representation compatible with the Hi-Tech PICC18 compiler.
//
// Returns
//    IEEE754 representation of a floating point no. packed into the lower
//    24 bits of an int.
//--------------------------------------------------------------------------
function FloatToIEEE754(Number : Double) : Integer;
var
  SignBit, Exponent : Integer;
  Mantissa : Double;
begin
  // Sign bit
  if Number < 0.0 then
  begin
    SignBit := 1;  // -ve
    Number := Number * (-1.0);
  end
  else
    SignBit := 0;  // +ve
  if Number <> 0 then
  begin
    // Exponent
    Exponent := Trunc(Log2(Number));
    if Exponent > Log2(Number) then
      Exponent := Exponent - 1;  // Make sure that 2^Exponent is less than Number
    // Mantissa
    Mantissa := (Number / IntPower(2, Exponent) - 1) * 32768.0;
    // Binary representation of floating point number
    Result := SignBit * 8388608 + (Exponent + 127) * 32768 + Round(Mantissa);
  end
  else
    Result := 0;
end;
//--------------------------------------------------------------------------
// IEEE754ToFloat
//
// Decode a floating point number encoded according to the truncated 24 bit
// IEEE754 representation compatible with the Hi-Tech PICC18 compiler.
//
// Returns
//    Float value decoded from IEEE754 binary representation
//--------------------------------------------------------------------------
function IEEE754ToFloat(Number : Integer) : Double;
var
  Exponent, Mantissa : Double;
begin
  if Number <> 0 then
  begin
    Exponent := IntPower(2, (((Number shr 15) and $FF) - 127));
    // Sign bit
    if (Number and $800000) > 0 then
      Exponent := Exponent * (-1.0);  // -ve value
    Mantissa := 1.0 + (Number and $7FFF) / 32768.0;
    Result := Exponent * Mantissa;
    if IsNaN(Result) then
      Result := NaN  // Not a number
    else
    if IsInfinite(Result) then
      Result := Infinity;  // Infinity
  end
  else
    Result := 0.0;
end;
//--------------------------------------------------------------------------
// CheckFloatingPointNumber
//
// Check that a string representing a floating point no. only contains the
// characters [0..9, '.', '-', 'E'] & is a valid float value.
//
// Returns
//    False = Invalid char. present
//    True = Valid floating point no.
//--------------------------------------------------------------------------
function CheckFloatingPointNumber(Number : String) : Boolean;
var
  CharNo : Integer;
  ValidNumber : Boolean;
  Value : Double;
begin
  if (Number <> '') then
  begin
    // Check string is a float
    ValidNumber := True;  // Assume valid
    for CharNo := 1 to Length(Number) do
    begin
      if not CharInSet(Number[CharNo], ['0'..'9', '.', '-', 'E']) then
        ValidNumber := False;
    end;
    if ValidNumber then
      ValidNumber := TryStrToFloat(Number, Value);
  end
  else
    ValidNumber := False;  // Invalid
  Result := ValidNumber;
end;
//--------------------------------------------------------------------------
end.
//--------------------------------------------------------------------------
//--------------------------------------------------------------------------

File: Driver/Source/IEEE754.c
//------------------------------------------------------------------------------
//  TITLE :- IEEE754 floating point number routines - IEEE754.c
//  AUTHOR :- Martyn Brown
//  DATE :- 19/05/14
//
//  DESCRIPTION :- Routines to deal with encoding/decoding of floating point
//             		 numbers in a binary format.
//
//  UPDATES :-
//  19/05/14 IEEE754.pas v4.0 converted to 'C' for Linux library
//
//  LICENSE :-
//  The CANdo API library source code is proprietary software & must not be
//  copied, modified or re-distributed without prior permission from the
//  author. The source code is provided to allow compilation of the CANdo API
//  library on any compatible platform. The resulting binary library file
//  'libCANdo.so' maybe freely copied & distributed without any restriction or
//  time limit, for personal, educational or commercial use. The software is
//  supplied "as is", with no implied warranties or guarantees.
//
//  (c) 2014 Netronics Ltd. All rights reserved.
//------------------------------------------------------------------------------
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include "SiUSBXp.h"
//--------------------------------------------------------------------------
// FloatToIEEE754
//
// Encode a floating point number according to the truncated 24 bit IEEE754
// representation compatible with the Hi-Tech PICC18 compiler.
//
// Returns
//    IEEE754 representation of a floating point no. packed into the lower
//    24 bits of an int.
//--------------------------------------------------------------------------
int FloatToIEEE754(double Number)
{
  int Result, SignBit, Exponent;
  double Mantissa;

  // Sign bit
  if (Number < 0.0)
  {
    SignBit = 1;  // -ve
    Number = Number * (-1.0);
  }
  else
    SignBit = 0;  // +ve
  if (Number != 0.0)
  {
    // Exponent
    Exponent = trunc(log2(Number));
    if (Exponent > log2(Number))
      Exponent = Exponent - 1;  // Make sure that 2^Exponent is less than Number
    // Mantissa
    Mantissa = (Number / exp2(Exponent) - 1) * 32768.0;
    // Binary representation of floating point number
    Result = SignBit * 8388608 + (Exponent + 127) * 32768 + round(Mantissa);
  }
  else
    Result = 0;

  return Result;
}
//--------------------------------------------------------------------------
// IEEE754ToFloat
//
// Decode a floating point number encoded according to the truncated 24 bit
// IEEE754 representation compatible with the Hi-Tech PICC18 compiler.
//
// Returns
//    Float value decoded from IEEE754 binary representation
//--------------------------------------------------------------------------
double IEEE754ToFloat(int Number)
{
  double Result, Exponent, Mantissa;

  if (Number != 0)
  {
    Exponent = exp2(((Number >> 15) & 0xFF) - 127);
    // Sign bit
    if ((Number & 0x800000) > 0)
      Exponent = Exponent * (-1.0);  // -ve value
    Mantissa = ((Number & 0x7FFF) / 32768) + 1.0;
    Result = Exponent * Mantissa;
    if (isnan(Result))
      Result = NAN;  // Not a number
    else
    if (!isfinite(Result))
      Result = INFINITY;  // Infinity
  }
  else
    Result = 0.0;

  return Result;
}
//--------------------------------------------------------------------------
// CheckFloatingPointNumber
//
// Check that a string representing a floating point no. only contains the
// characters [0..9, '.', '-', 'e', 'E'] & is a valid float value.
//
// Returns
//    False = Invalid char. present
//    True = Valid floating point no.
//--------------------------------------------------------------------------
unsigned char CheckFloatingPointNumber(const char * Number)
{
  unsigned char ValidNumberFlag;
  double Value;

  if (strcmp(Number, "") != 0)
  {
    // Check string is a float
  	Value = strtod(Number, NULL);
  	if (Value == 0.0)
  		ValidNumberFlag = FALSE;  // Invalid
  	else
  		ValidNumberFlag = TRUE;  // Valid
  }
  else
    ValidNumberFlag = FALSE;  // Invalid

  return ValidNumberFlag;
}
//------------------------------------------------------------------------------
//------------------------------------------------------------------------------
